Mathlib/CategoryTheory/Limits/Shapes/FunctorToTypes.lean

```diff
@@ -0,0 +1,290 @@
+    z = w := Prod.ext h1 h2
+
+variable (F G)
+
+/-- `(F ⨯ G).obj a` is in bijection with the product of `F.obj a` and `G.obj a`. -/
+@[simps]
+noncomputable
+def binaryProductEquiv (a : C) : (F ⨯ G).obj a ≃ (F.obj a) × (G.obj a) where
+  toFun z := ⟨((binaryProductIso F G).hom.app a z).1, ((binaryProductIso F G).hom.app a z).2⟩
+  invFun z := prodMk z.1 z.2
+  left_inv _ := by simp [prodMk]
+  right_inv _ := by simp [prodMk]
+
+@[ext]
+lemma prod_ext' (a : C) (z w : (F ⨯ G).obj a)
+    (h1 : (Limits.prod.fst (X := F)).app a z = (Limits.prod.fst (X := F)).app a w)
+    (h2 : (Limits.prod.snd (X := F)).app a z = (Limits.prod.snd (X := F)).app a w) :
+    z = w := by
+  apply Equiv.injective (binaryProductEquiv F G a)
+  aesop
+
+end prod
+
+section coprod
+
+/-- `coprod F G` is the explicit binary coproduct of type-valued functors `F` and `G`. -/
+def coprod : C ⥤ Type w where
+  obj a := F.obj a ⊕ G.obj a
+  map f x := by
+    cases x with
+    | inl x => exact .inl (F.map f x)
+    | inr x => exact .inr (G.map f x)
+
+variable {F G}
+
+/-- The left inclusion of `F` into `coprod F G`. -/
+@[simps]
+def coprod.inl : F ⟶ coprod F G where
+  app _ x := .inl x
+
+  /-- The right inclusion of `G` into `coprod F G`. -/
+@[simps]
+def coprod.inr : G ⟶ coprod F G where
+  app _ x := .inr x
+
+/-- Given natural transformations `F₁ ⟶ F` and `F₂ ⟶ F`, construct
+a natural transformation `coprod F₁ F₂ ⟶ F`. -/
+@[simps]
+def coprod.desc {F₁ F₂ : C ⥤ Type w} (τ₁ : F₁ ⟶ F) (τ₂ : F₂ ⟶ F) :
+    coprod F₁ F₂ ⟶ F where
+  app a x := by
+     cases x with
+     | inl x => exact τ₁.app a x
+     | inr x => exact τ₂.app a x
+  naturality _ _ _:= by
+    ext x
+    cases x with | _ => simp only [coprod, types_comp_apply, FunctorToTypes.naturality]
+
+@[simp]
+lemma coprod.desc_inl {F₁ F₂ : C ⥤ Type w} (τ₁ : F₁ ⟶ F) (τ₂ : F₂ ⟶ F) :
+    coprod.inl ≫ coprod.desc τ₁ τ₂ = τ₁ := rfl
+
+@[simp]
+lemma coprod.desc_inr {F₁ F₂ : C ⥤ Type w} (τ₁ : F₁ ⟶ F) (τ₂ : F₂ ⟶ F) :
+    coprod.inr ≫ coprod.desc τ₁ τ₂ = τ₂ := rfl
+
+variable (F G)
+
+/-- The binary cofan whose point is `coprod F G`. -/
+@[simps!]
+def binaryCoproductCocone : BinaryCofan F G :=
+  BinaryCofan.mk coprod.inl coprod.inr
+
+/-- `coprod F G` is a colimit cocone. -/
+@[simps]
+def binaryCoproductColimit : IsColimit (binaryCoproductCocone F G) where
+  desc (s : BinaryCofan F G) := coprod.desc s.inl s.inr
+  fac _ := fun ⟨j⟩ ↦ WalkingPair.casesOn j rfl rfl
+  uniq _ _ h := by
+    ext _ x
+    cases x with | _ => simp [← h ⟨WalkingPair.right⟩, ← h ⟨WalkingPair.left⟩]
+
+/-- `coprod F G` is a binary coproduct for `F` and `G`. -/
+def binaryCoproductColimitCocone : Limits.ColimitCocone (pair F G) :=
+  ⟨_, binaryCoproductColimit F G⟩
+
+/-- The categorical binary coproduct of type-valued functors is `coprod F G`. -/
+noncomputable def binaryCoproductIso : F ⨿ G ≅ coprod F G :=
+  colimit.isoColimitCocone (binaryCoproductColimitCocone F G)
+
+@[simp]
+lemma binaryCoproductIso_inl_comp_hom :
```

Comment:
This name would be better:
```suggestion
lemma inl_comp_binaryCoproductIso_hom :
```
(Similar names below should be changed accordingly.)